protocol/contracts/beanstalk/sun/SeasonFacet/Weather.sol

```diff
@@ -224,27 +303,106 @@ contract Weather is Sun {
-            address(sopToken),
-            amountOut,
-            newHarvestable
-        );
+        rewardSop(wellDeltaB.well, amountOut);
+        emit SeasonOfPlentyWell(s.season.current, wellDeltaB.well, address(sopToken), amountOut);
     }
 
     /**
      * @dev Allocate `sop token` during a Season of Plenty.
      */
-    function rewardSop(uint256 amount) private {
-        s.sops[s.season.rainStart] = s.sops[s.season.lastSop].add(
+    function rewardSop(address well, uint256 amount) private {
+        s.sops[s.season.rainStart][well] = s.sops[s.season.lastSop][well].add(
             amount.mul(C.SOP_PRECISION).div(s.r.roots)
         );
         s.season.lastSop = s.season.rainStart;
         s.season.lastSopSeason = s.season.current;
     }
 
+    /*
+     * @notice Calculates the amount of beans per well that should be minted in a sop.
+     * @param wellDeltaBs The deltaBs of all whitelisted wells in which to flood. Must be sorted in descending order.
+     */
+    // code review note: have some casual comments here for just understanding the code, happy to
+    // change this function to make it more efficient based on feedback.
+    function calculateSopPerWell(
+        WellDeltaB[] memory wellDeltaBs
+    ) public pure returns (WellDeltaB[] memory) {
+        uint256 totalPositiveDeltaB = 0;
+        uint256 totalNegativeDeltaB = 0;
+        uint256 positiveDeltaBCount = 0;
+
+        for (uint256 i = 0; i < wellDeltaBs.length; i++) {
+            if (wellDeltaBs[i].deltaB > 0) {
+                totalPositiveDeltaB += uint256(wellDeltaBs[i].deltaB);
+                positiveDeltaBCount++;
+            } else {
+                totalNegativeDeltaB += uint256(-wellDeltaBs[i].deltaB);
+            }
+        }
+
+        // most likely case is that all deltaBs are positive
+        if (positiveDeltaBCount == wellDeltaBs.length) {
+            // if all deltaBs are positive, need to sop all to zero
+            for (uint256 i = 0; i < wellDeltaBs.length; i++) {
+                wellDeltaBs[i].reductionAmount = uint256(wellDeltaBs[i].deltaB);
+            }
+            return wellDeltaBs;
+        }
+
+        if (positiveDeltaBCount == 0) {
+            // No positive values, so no well flooding needed
+            return wellDeltaBs;
+        }
+
+        if (totalPositiveDeltaB < totalNegativeDeltaB) {
+            // in theory this should never happen because overall deltaB is required for sop
+            revert("Flood: Overall deltaB is negative");
+        }
+
+        uint256 shaveOff = totalPositiveDeltaB - totalNegativeDeltaB;
+        uint256 cumulativeTotal;
+        uint256 shaveToLevel;
+
+        for (uint256 i = 0; i <= positiveDeltaBCount; i++) {
+            if (positiveDeltaBCount == 1 || i == 0) {
+                // reduce enough to equal the negative deltaB
+                if (positiveDeltaBCount == 1) {
+                    wellDeltaBs[i].reductionAmount = uint256(wellDeltaBs[i].deltaB).sub(
```

Comment:
Yes, otherwise it'll hit the `if (totalPositiveDeltaB < totalNegativeDeltaB)` statement above, wrote a test here to verify this works as expected. Lmk if this is what you were thinking, or if there's another possible way to break this?

https://github.com/BeanstalkFarms/Beanstalk/commit/4bb95df9fb033bd7da5ae1b3860a0a97e7d83b83